Make `azure_keyvault_backend()` handle DNS resolution issues

# The what

4 days ago, the nightly CI started failing. This is likely due to DNS changes for the domain passed in tests — `test.vault.azure.net` returns `NXDOMAIN` now but I imagine a few days ago it wasn't. A few "neighboring" subdomains still work and could be a short-term hack:

<details>
<summary>👉 DNS lookups (click to expand) 👈</summary>

```console
$ dig @1.1.1.1 test.vault.azure.net  

; <<>> DiG 9.18.42 <<>> @1.1.1.1 test.vault.azure.net
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 59603
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;test.vault.azure.net.		IN	A

;; AUTHORITY SECTION:
vault.azure.net.	60	IN	SOA	ns1-01.azure-dns.com. msnhst.microsoft.com. 10001 900 300 604800 60

;; Query time: 16 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Tue Jan 13 18:05:54 CET 2026
;; MSG SIZE  rcvd: 122

$ dig @1.1.1.1 test42.vault.azure.net

; <<>> DiG 9.18.42 <<>> @1.1.1.1 test42.vault.azure.net
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55402
;; flags: qr rd ra; QUERY: 1, ANSWER: 8, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;test42.vault.azure.net.		IN	A

;; ANSWER SECTION:
test42.vault.azure.net.	60	IN	CNAME	test42.privatelink.vaultcore.azure.net.
test42.privatelink.vaultcore.azure.net.	60 IN CNAME data-prod-dewc.vaultcore.azure.net.
data-prod-dewc.vaultcore.azure.net. 60 IN CNAME	data-prod-dewc-region.vaultcore.azure.net.
data-prod-dewc-region.vaultcore.azure.net. 60 IN CNAME dwc.prd.r.kv.aadg.msidentity.com.
dwc.prd.r.kv.aadg.msidentity.com. 300 IN CNAME	dwc.tm.prd.r.kv.aadg.trafficmanager.net.
dwc.tm.prd.r.kv.aadg.trafficmanager.net. 300 IN	A 51.116.154.64
dwc.tm.prd.r.kv.aadg.trafficmanager.net. 300 IN	A 51.116.243.222
dwc.tm.prd.r.kv.aadg.trafficmanager.net. 300 IN	A 51.116.251.190

;; Query time: 79 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Tue Jan 13 18:02:03 CET 2026
;; MSG SIZE  rcvd: 303
```
</details>

Last successful nightly (Jan 8, 2026 @ 06:12 AM GMT+1): https://github.com/ansible/awx-plugins/actions/runs/20806290240
First failing nightly (Jan 9, 2026 @ 06:12 AM GMT+1): https://github.com/ansible/awx-plugins/actions/runs/20841988704

<details>
<summary>👉 Relevant trace excerpt for history (click to expand) 👈</summary>

```python-traceback
__________________________ test_azure_kv_invalid_env ___________________________

    def test_azure_kv_invalid_env() -> None:
        """Test running outside of Azure raises error."""
        error_msg = (
            'You are not operating on an Azure VM, so the Managed Identity '
            'feature is unavailable. Please provide the full Client ID, '
            'Client Secret, and Tenant ID or run the software on an Azure VM.'
        )
    
        with pytest.raises(
            RuntimeError,
            match=error_msg,
        ):
>           azure_kv.azure_keyvault_backend(
                url='https://test.vault.azure.net',
                client='',
                secret='client-secret',
                tenant='tenant-id',
                secret_field='secret',
                secret_version='',
            )

error_msg  = ('You are not operating on an Azure VM, so the Managed Identity feature is '
 'unavailable. Please provide the full Client ID, Client Secret, and Tenant ID '
 'or run the software on an Azure VM.')

tests/azure_kv_test.py:37: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/python3.14/site-packages/awx_plugins/credentials/azure_kv.py:149: in azure_keyvault_backend
    keyvault_secret = keyvault.get_secret(
        chosen_credential = <azure.identity._credentials.managed_identity.ManagedIdentityCredential object at 0x1063ee900>
        client     = ''
        keyvault   = <azure.keyvault.secrets._client.SecretClient object at 0x106521160>
        secret     = 'client-secret'
        secret_field = 'secret'
        secret_version = ''
        tenant     = 'tenant-id'
        url        = 'https://test.vault.azure.net'
lib/python3.14/site-packages/azure/core/tracing/decorator.py:119: in wrapper_use_tracer
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
        args       = (<azure.keyvault.secrets._client.SecretClient object at 0x106521160>,)
        func       = <function SecretClient.get_secret at 0x105b04930>
        kind       = <SpanKind.INTERNAL: 6>
        kwargs     = {'name': 'secret', 'version': ''}
        merge_span = False
        name_of_span = None
        passed_in_parent = None
        tracing_attributes = {}
        tracing_enabled = False
        tracing_options = {}
        user_enabled = None
lib/python3.14/site-packages/azure/keyvault/secrets/_client.py:65: in get_secret
    bundle = self._client.get_secret(
        kwargs     = {}
        name       = 'secret'
        self       = <azure.keyvault.secrets._client.SecretClient object at 0x106521160>
        version    = ''
lib/python3.14/site-packages/azure/core/tracing/decorator.py:119: in wrapper_use_tracer
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
        args       = (<azure.keyvault.secrets._generated._client.KeyVaultClient object at 0x106521400>,)
        func       = <function KeyVaultClientOperationsMixin.get_secret at 0x105a73480>
        kind       = <SpanKind.INTERNAL: 6>
        kwargs     = {'secret_name': 'secret', 'secret_version': ''}
        merge_span = False
        name_of_span = None
        passed_in_parent = None
        tracing_attributes = {}
        tracing_enabled = False
        tracing_options = {}
        user_enabled = None
lib/python3.14/site-packages/azure/keyvault/secrets/_generated/_operations/_operations.py:802: in get_secret
    pipeline_response: PipelineResponse = self._client._pipeline.run(  # pylint: disable=protected-access
        _headers   = {}
        _params    = {}
        _request   = <HttpRequest [GET], url: 'https://test.vault.azure.net/secrets/secret/?api-version=7.6'>
        _stream    = False
        cls        = None
        error_map  = {304: <class 'azure.core.exceptions.ResourceNotModifiedError'>,
 401: <class 'azure.core.exceptions.ClientAuthenticationError'>,
 404: <class 'azure.core.exceptions.ResourceNotFoundError'>,
 409: <class 'azure.core.exceptions.ResourceExistsError'>}
        kwargs     = {}
        path_format_arguments = {'vaultBaseUrl': 'https://test.vault.azure.net'}
        secret_name = 'secret'
        secret_version = ''
        self       = <azure.keyvault.secrets._generated._client.KeyVaultClient object at 0x106521400>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:242: in run
    return first_node.send(pipeline_request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        context    = {'logger': <Logger azure.core.pipeline.policies.http_logging_policy (WARNING)>,
 'logging_enable': False,
 'retry_count': 3}
        first_node = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1064f9550>
        kwargs     = {'stream': False}
        pipeline_request = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        request    = <HttpRequest [GET], url: 'https://test.vault.azure.net/secrets/secret/?api-version=7.6'>
        self       = <azure.core.pipeline._base.Pipeline object at 0x1064239d0>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1064f9550>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1064f9950>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x10645ac60>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x10645ad50>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1064ebe70>
lib/python3.14/site-packages/azure/core/pipeline/policies/_redirect.py:205: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        original_domain = 'test.vault.azure.net'
        redirect_settings = {'allow': True, 'history': [], 'redirects': 30}
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        retryable  = True
        self       = <azure.core.pipeline.policies._redirect.RedirectPolicy object at 0x1065216a0>
lib/python3.14/site-packages/azure/core/pipeline/policies/_retry.py:567: in send
    raise err
        absolute_timeout = 604795.043141365
        end_time   = 1767935683.1784499
        is_response_error = False
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        response   = None
        retry_active = False
        retry_settings = {'backoff': 0.8,
 'body_position': None,
 'connect': -1,
 'file_positions': None,
 'history': [<azure.core.pipeline.policies._base.RequestHistory object at 0x106523620>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106539310>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106539a90>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106554d60>],
 'max_backoff': 120,
 'methods': frozenset({'DELETE', 'HEAD', 'PUT', 'OPTIONS', 'GET', 'TRACE'}),
 'read': 3,
 'status': 3,
 'timeout': 604800,
 'total': 6}
        self       = <azure.core.pipeline.policies._retry.RetryPolicy object at 0x1065217f0>
        start_time = 1767935683.1726131
        transport  = <azure.core.pipeline.transport._requests_basic.RequestsTransport object at 0x106423750>
lib/python3.14/site-packages/azure/core/pipeline/policies/_retry.py:545: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        absolute_timeout = 604795.043141365
        end_time   = 1767935683.1784499
        is_response_error = False
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        response   = None
        retry_active = False
        retry_settings = {'backoff': 0.8,
 'body_position': None,
 'connect': -1,
 'file_positions': None,
 'history': [<azure.core.pipeline.policies._base.RequestHistory object at 0x106523620>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106539310>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106539a90>,
             <azure.core.pipeline.policies._base.RequestHistory object at 0x106554d60>],
 'max_backoff': 120,
 'methods': frozenset({'DELETE', 'HEAD', 'PUT', 'OPTIONS', 'GET', 'TRACE'}),
 'read': 3,
 'status': 3,
 'timeout': 604800,
 'total': 6}
        self       = <azure.core.pipeline.policies._retry.RetryPolicy object at 0x1065217f0>
        start_time = 1767935683.1726131
        transport  = <azure.core.pipeline.transport._requests_basic.RequestsTransport object at 0x106423750>
lib/python3.14/site-packages/azure/keyvault/secrets/_shared/challenge_auth_policy.py:113: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.keyvault.secrets._shared.challenge_auth_policy.ChallengeAuthPolicy object at 0x1065212b0>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x106528050>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1064f4870>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1063cfb90>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x1063cfd10>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:98: in send
    response = self.next.send(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._SansIOHTTPPolicyRunner object at 0x106512990>
lib/python3.14/site-packages/azure/core/pipeline/_base.py:130: in send
    self._sender.send(request.http_request, **request.context.options),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        request    = <azure.core.pipeline.PipelineRequest object at 0x106521e80>
        self       = <azure.core.pipeline._base._TransportRunner object at 0x106423b10>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <azure.core.pipeline.transport._requests_basic.RequestsTransport object at 0x106423750>
request = <HttpRequest [GET], url: 'https://test.vault.azure.net/secrets/secret/?api-version=7.6'>
proxies = None, kwargs = {'stream': False}, response = None
error = ServiceRequestError("<urllib3.connection.HTTPSConnection object at 0x106539bd0>: Failed to resolve 'test.vault.azure.net' ([Errno 8] nodename nor servname provided, or not known)")
connection_timeout = 300, timeout = (300, 300), read_timeout = 300

    def send(  # pylint: disable=too-many-statements
        self,
        request: Union[HttpRequest, "RestHttpRequest"],
        *,
        proxies: Optional[MutableMapping[str, str]] = None,
        **kwargs
    ) -> Union[HttpResponse, "RestHttpResponse"]:
        """Send request object according to configuration.
    
        :param request: The request object to be sent.
        :type request: ~azure.core.pipeline.transport.HttpRequest
        :return: An HTTPResponse object.
        :rtype: ~azure.core.pipeline.transport.HttpResponse
    
        :keyword MutableMapping proxies: will define the proxy to use. Proxy is a dict (protocol, url)
        """
        self.open()
        response = None
        error: Optional[AzureErrorUnion] = None
    
        try:
            connection_timeout = kwargs.pop("connection_timeout", self.connection_config.timeout)
    
            if isinstance(connection_timeout, tuple):
                if "read_timeout" in kwargs:
                    raise ValueError("Cannot set tuple connection_timeout and read_timeout together")
                _LOGGER.warning("Tuple timeout setting is deprecated")
                timeout = connection_timeout
            else:
                read_timeout = kwargs.pop("read_timeout", self.connection_config.read_timeout)
                timeout = (connection_timeout, read_timeout)
            response = self.session.request(  # type: ignore
                request.method,
                request.url,
                headers=request.headers,
                data=request.data,
                files=request.files,
                verify=kwargs.pop("connection_verify", self.connection_config.verify),
                timeout=timeout,
                cert=kwargs.pop("connection_cert", self.connection_config.cert),
                allow_redirects=False,
                proxies=proxies,
                **kwargs
            )
            response.raw.enforce_content_length = True
    
        except AttributeError as err:
            if self.session is None:
                raise ValueError(
                    "No session available for request. "
                    "Please report this issue to https://github.com/Azure/azure-sdk-for-python/issues."
                ) from err
            raise
        except (
            NewConnectionError,
            ConnectTimeoutError,
        ) as err:
            error = ServiceRequestError(err, error=err)
        except requests.exceptions.ReadTimeout as err:
            error = ServiceResponseError(err, error=err)
        except requests.exceptions.ConnectionError as err:
            if err.args and isinstance(err.args[0], ProtocolError):
                error = ServiceResponseError(err, error=err)
            else:
                error = ServiceRequestError(err, error=err)
        except requests.exceptions.ChunkedEncodingError as err:
            msg = err.__str__()
            if "IncompleteRead" in msg:
                _LOGGER.warning("Incomplete download.")
                error = IncompleteReadError(err, error=err)
            else:
                _LOGGER.warning("Unable to stream download.")
                error = HttpResponseError(err, error=err)
        except requests.RequestException as err:
            error = ServiceRequestError(err, error=err)
    
        if error:
>           raise error
E           azure.core.exceptions.ServiceRequestError: <urllib3.connection.HTTPSConnection object at 0x106539bd0>: Failed to resolve 'test.vault.azure.net' ([Errno 8] nodename nor servname provided, or not known)

connection_timeout = 300
error      = ServiceRequestError("<urllib3.connection.HTTPSConnection object at 0x106539bd0>: Failed to resolve 'test.vault.azure.net' ([Errno 8] nodename nor servname provided, or not known)")
kwargs     = {'stream': False}
proxies    = None
read_timeout = 300
request    = <HttpRequest [GET], url: 'https://test.vault.azure.net/secrets/secret/?api-version=7.6'>
response   = None
self       = <azure.core.pipeline.transport._requests_basic.RequestsTransport object at 0x106423750>
timeout    = (300, 300)

lib/python3.14/site-packages/azure/core/pipeline/transport/_requests_basic.py:411: ServiceRequestError
```
</details>

# What's next

The action items would be:
* [ ] Improve `azure_keyvault_backend()` to also handle `azure.core.exceptions.ServiceRequestError` or its base exception, converting it into a `RuntimeError`
* [ ] Make the tests stop relying on the outside world — [RFC 2606](https://www.rfc-editor.org/rfc/rfc2606.html) suggests a few reserved domains for use in testing

# Extra context

Exception handling and tests in question were added in #106 which may hold additional discussions around the current design.

--- a/src/awx_plugins/credentials/azure_kv.py
+++ b/src/awx_plugins/credentials/azure_kv.py
@@ -19,6 +19,7 @@ from awx_plugins.interfaces._temporary_private_django_api import (  # noqa: WPS4
 )
 
 from azure.core.credentials import TokenCredential
+from azure.core.exceptions import ServiceRequestError
 from azure.identity import (
     ClientSecretCredential,
     CredentialUnavailableError,
@@ -157,6 +158,10 @@ def azure_keyvault_backend(  # noqa: WPS211
             'feature is unavailable. Please provide the full Client ID, '
             'Client Secret, and Tenant ID or run the software on an Azure VM.',
         ) from secret_lookup_err
+    except ServiceRequestError as request_err:
+        raise RuntimeError(
+            f'Failed to connect to Azure Key Vault: {request_err}',
+        ) from request_err
     return keyvault_secret.value
 
 

--- a/src/awx_plugins/credentials/azure_kv.py
+++ b/src/awx_plugins/credentials/azure_kv.py
@@ -164,7 +164,8 @@ def azure_keyvault_backend(  # noqa: WPS211
         ) from request_err
     except _az_exc.AzureError as catchall_azure_error:
         raise RuntimeError(
-            f'Error retrieving secret from Azure Key Vault: {catchall_azure_error}',
+            'Error retrieving secret from Azure Key Vault: '
+            f'{catchall_azure_error}',
         ) from catchall_azure_error
     return keyvault_secret.value
 

--- a/src/awx_plugins/credentials/azure_kv.py
+++ b/src/awx_plugins/credentials/azure_kv.py
@@ -19,7 +19,7 @@ from awx_plugins.interfaces._temporary_private_django_api import (  # noqa: WPS4
 )
 
 from azure.core.credentials import TokenCredential
-from azure.core.exceptions import ServiceRequestError
+from azure.core.exceptions import AzureError, ServiceRequestError
 from azure.identity import (
     ClientSecretCredential,
     CredentialUnavailableError,
@@ -162,6 +162,10 @@ def azure_keyvault_backend(  # noqa: WPS211
         raise RuntimeError(
             f'Failed to connect to Azure Key Vault: {request_err}',
         ) from request_err
+    except AzureError as error:
+        raise RuntimeError(
+            f'Error retrieving secret from Azure Key Vault: {error}',
+        )
     return keyvault_secret.value
 
 

--- a/src/awx_plugins/credentials/azure_kv.py
+++ b/src/awx_plugins/credentials/azure_kv.py
@@ -18,8 +18,8 @@ from awx_plugins.interfaces._temporary_private_django_api import (  # noqa: WPS4
     gettext_noop as _,
 )
 
+from azure.core import exceptions as _az_exc
 from azure.core.credentials import TokenCredential
-from azure.core.exceptions import AzureError, ServiceRequestError
 from azure.identity import (
     ClientSecretCredential,
     CredentialUnavailableError,
@@ -158,14 +158,14 @@ def azure_keyvault_backend(  # noqa: WPS211
             'feature is unavailable. Please provide the full Client ID, '
             'Client Secret, and Tenant ID or run the software on an Azure VM.',
         ) from secret_lookup_err
-    except ServiceRequestError as request_err:
+    except _az_exc.ServiceRequestError as request_err:
         raise RuntimeError(
             f'Failed to connect to Azure Key Vault: {request_err}',
         ) from request_err
-    except AzureError as error:
+    except _az_exc.AzureError as catchall_azure_error:
         raise RuntimeError(
-            f'Error retrieving secret from Azure Key Vault: {error}',
-        )
+            f'Error retrieving secret from Azure Key Vault: {catchall_azure_error}',
+        ) from catchall_azure_error
     return keyvault_secret.value
 
 

--- a/tests/azure_kv_test.py
+++ b/tests/azure_kv_test.py
@@ -37,13 +37,13 @@ def test_azure_kv_invalid_env(
     simulate the expected CredentialUnavailableError.
     """
     mock_client = mocker.Mock()
+    mock_client = mocker.patch.object(azure_kv, 'SecretClient')
     mock_client.return_value.get_secret.side_effect = (
         CredentialUnavailableError(
             message='ManagedIdentityCredential authentication unavailable.',
         )
     )
     monkeypatch.setattr(azure_kv, 'SecretClient', mock_client)
-
     error_msg = (
         'You are not operating on an Azure VM, so the Managed Identity '
         'feature is unavailable. Please provide the full Client ID, '
@@ -55,8 +55,24 @@ def test_azure_kv_invalid_env(
         match=error_msg,
     ):
         azure_kv.azure_keyvault_backend(
-            url='https://test.vault.azure.net',
+            url='https://keyvault.test',
             client='',
+            secret='client-secret',
+            tenant='tenant-id',
+            secret_field='secret',
+            secret_version='',
+        )
+
+
+def test_azure_kv_dns_error() -> None:
+    """Test DNS resolution error is converted to RuntimeError."""
+    with pytest.raises(
+        RuntimeError,
+        match='Failed to connect to Azure Key Vault',
+    ):
+        azure_kv.azure_keyvault_backend(
+            url='https://keyvault.test',
+            client='client-id',
             secret='client-secret',
             tenant='tenant-id',
             secret_field='secret',
@@ -90,7 +106,7 @@ def test_azure_kv_valid_auth(
     )
 
     keyvault_secret = azure_kv.azure_keyvault_backend(
-        url='https://test.vault.azure.net',
+        url='https://keyvault.test',
         client=client,
         secret=secret,
         tenant=tenant,

--- a/tests/azure_kv_test.py
+++ b/tests/azure_kv_test.py
@@ -3,6 +3,7 @@
 import pytest
 from pytest_mock import MockerFixture
 
+from azure.core.exceptions import AzureError
 from azure.identity import CredentialUnavailableError
 from azure.keyvault.secrets import (
     KeyVaultSecret,
@@ -69,6 +70,32 @@ def test_azure_kv_dns_error() -> None:
     with pytest.raises(
         RuntimeError,
         match='Failed to connect to Azure Key Vault',
+    ):
+        azure_kv.azure_keyvault_backend(
+            url='https://keyvault.test',
+            client='client-id',
+            secret='client-secret',
+            tenant='tenant-id',
+            secret_field='secret',
+            secret_version='',
+        )
+
+
+def test_azure_kv_generic_azure_error(
+    mocker: MockerFixture,
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Test generic AzureError is converted to RuntimeError."""
+    mock_client = mocker.Mock()
+    mock_client = mocker.patch.object(azure_kv, 'SecretClient')
+    mock_client.return_value.get_secret.side_effect = AzureError(
+        message='Secret not found or access denied',
+    )
+    monkeypatch.setattr(azure_kv, 'SecretClient', mock_client)
+
+    with pytest.raises(
+        RuntimeError,
+        match='Error retrieving secret from Azure Key Vault',
     ):
         azure_kv.azure_keyvault_backend(
             url='https://keyvault.test',

--- a/tests/azure_kv_test.py
+++ b/tests/azure_kv_test.py
@@ -27,7 +27,6 @@ class _FakeSecretClient(SecretClient):
 
 def test_azure_kv_invalid_env(
     mocker: MockerFixture,
-    monkeypatch: pytest.MonkeyPatch,
 ) -> None:
     """Test running outside of Azure raises error.
 
@@ -37,14 +36,13 @@ def test_azure_kv_invalid_env(
     a DNS error before the credential check, so we mock SecretClient to
     simulate the expected CredentialUnavailableError.
     """
-    mock_client = mocker.Mock()
-    mock_client = mocker.patch.object(azure_kv, 'SecretClient')
+    mock_client = mocker.patch.object(azure_kv, 'SecretClient', autospec=True)
     mock_client.return_value.get_secret.side_effect = (
         CredentialUnavailableError(
             message='ManagedIdentityCredential authentication unavailable.',
         )
     )
-    monkeypatch.setattr(azure_kv, 'SecretClient', mock_client)
+
     error_msg = (
         'You are not operating on an Azure VM, so the Managed Identity '
         'feature is unavailable. Please provide the full Client ID, '
@@ -69,7 +67,7 @@ def test_azure_kv_dns_error() -> None:
     """Test DNS resolution error is converted to RuntimeError."""
     with pytest.raises(
         RuntimeError,
-        match='Failed to connect to Azure Key Vault',
+        match=r'^Failed to connect to Azure Key Vault: .',
     ):
         azure_kv.azure_keyvault_backend(
             url='https://keyvault.test',
@@ -83,19 +81,16 @@ def test_azure_kv_dns_error() -> None:
 
 def test_azure_kv_generic_azure_error(
     mocker: MockerFixture,
-    monkeypatch: pytest.MonkeyPatch,
 ) -> None:
     """Test generic AzureError is converted to RuntimeError."""
-    mock_client = mocker.Mock()
-    mock_client = mocker.patch.object(azure_kv, 'SecretClient')
+    mock_client = mocker.patch.object(azure_kv, 'SecretClient', autospec=True)
     mock_client.return_value.get_secret.side_effect = AzureError(
         message='Secret not found or access denied',
     )
-    monkeypatch.setattr(azure_kv, 'SecretClient', mock_client)
 
     with pytest.raises(
         RuntimeError,
-        match='Error retrieving secret from Azure Key Vault',
+        match=r'^Error retrieving secret from Azure Key Vault: .',
     ):
         azure_kv.azure_keyvault_backend(
             url='https://keyvault.test',

